src/endpoint/s3/s3_rest.js

```diff
@@ -231,63 +231,70 @@ function authenticate_request(req, res) {
     }
 }
 
-function parse_op_name(req) {
-    const virtual_host_suffix = req.virtual_host_suffix;
-    const method = req.method.toLowerCase();
-    const host = req.headers.host.split(':')[0]; // cutting off port
-    const url = req.url;
-
-    var bucket = '';
-    var key = '';
-    var i;
+function parse_bucket_and_key(req) {
+    const { url, headers, virtual_hosts } = req;
+    const host = headers.host.split(':')[0]; // cutting off port
 
     // see http://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html
-    const is_path_style = !host ||
-        !virtual_host_suffix || // added case for when no DNS name is defined - if we work with IP always using path-style
-        virtual_host_suffix === '.' + host ||
+    const is_path_style =
+        !host ||
         net.isIP(host) ||
-        host === 'localhost'; // we added this case on top of the S3 doc cases to handle requests with IP host
-
-    const is_bucket_subdomain =
-        virtual_host_suffix &&
-        host.endsWith(virtual_host_suffix) &&
-        host !== virtual_host_suffix; // non empty bucket name
+        host === 'localhost' || // we added this case on top of the S3 doc cases to handle requests with IP host
+        virtual_hosts.length === 0 ||
+        virtual_hosts.some(virtual_host => virtual_host === host);
 
+    let bucket;
+    let key;
+    let is_virtual_host_bucket;
     if (is_path_style) {
         const index = url.indexOf('/', 1);
         const pos = index < 0 ? url.length : index;
+
         bucket = url.slice(1, pos);
         key = url.slice(pos + 1);
-    } else if (is_bucket_subdomain) {
-        bucket = host.slice(0, -virtual_host_suffix.length);
-        key = url.slice(1);
-        req.virtual_hosted_bucket = bucket;
-    } else { // bucket is host - assume DNS points that name to us
-        bucket = host;
+        is_virtual_host_bucket = false;
+
+    } else {
+        const match = virtual_hosts
+            .map(virtual_host => `.${virtual_host}`)
+            .find(suffix => host.endsWith(suffix));
```

Comment:
@guymguym Array.map and Array.find are not actual calls to some native/internal functions in V8. 
The actual generated V8 op code will remove and inline both this function into something more in the like of the loop you describe. 
On the other hand for of loop are more expansive because of the iterator result that is generated for every item (maybe in the future the V8 theme will optimize the "for of" for simple arrays to skip the iterator all together).

Regarding the code readability, I always prefer to code more functional than imperative over arrays (map, filter, reduce, fund, etc.) because you don't need to specify how to loop but you want to achieve and in my eyes it is even more readable.

saying all that the agree that the map is redundant and i will change it to:
```
const match = virtual_hosts.find(virtual_host => host.endsWith(`.${virtual_host}`));
```
which keep match a const and not a var (or a let), this will give the compiler more opportunities for optimizations, like not allocating a frame variable the const

Reply:
ok